### DISCUSSION TITLE
How to invoke a function that returns a C# object?

### DISCUSSION THREAD

**Discussion by lkwargs** [vfsfitvnm/frida-il2cpp-bridge#311] [Q&A]
I want to hook some get/post functions in an unity game app, and I write the defination of these functions in my script.  The problem is the app stucked after I invoke the original function. I guess I use the wrong way to invoke the function.
Also, I tried to use the [script](https://github.com/vfsfitvnm/frida-il2cpp-bridge/wiki/Snippets#tracing) to trace all methods calls of System.String, but the app also stucked at the same time.
I tried my best to looking for templates and answers in the wiki and discussion, but I still cannot find the solution. Here is my script:
```ts
import "frida-il2cpp-bridge";

Il2Cpp.perform(() => {
    const SystemString = Il2Cpp.corlib.class("System.String");
    const assemblyCSharp = Il2Cpp.domain.assembly("Assembly-CSharp").image;
    const luaSender = assemblyCSharp.class("Torappu.Lua.LuaSender");
    const networker = assemblyCSharp.class("Torappu.Network.Networker");
    const luaSendGet = luaSender.method("SendGet");
    const sendGet = networker.method("SendGet");
    const sendPost = networker.method("SendPost");
    const sendPost2 = sendPost.overload("System.String", "System.String", "System.String", "System.Collections.Generic.Dictionary<System.String,System.String>");

    // static System.String SendGet(System.String url, System.String param, System.Boolean useLoadingMask); // 0x011b30ca
    luaSendGet.implementation = function(url: Il2Cpp.String, param: Il2Cpp.String, useLoadingMask: UInt64) : Il2Cpp.String {
        // OnEnter
        console.log("luaSendGet");
        console.log(url.toString());
        console.log(param.toString());
        console.log(Boolean(useLoadingMask));
        const retValue = this.method<Il2Cpp.String>("SendGet").invoke(url, param, useLoadingMask);
        // OnLeave
        console.log(retValue.toString());
        return retValue as Il2Cpp.String;
    }

    // Torappu.Network.WebHttpResult SendGet(System.String url, System.String param); // 0x02367f2a
    sendGet.implementation = function(url: Il2Cpp.String, param: Il2Cpp.String) : Il2Cpp.Object {
        // OnEnter
        console.log("sendGet");
        console.log(url.toString());
        console.log(param.toString());
        const retValue = this.method<Il2Cpp.Object>("SendGet").invoke(url, param);
        // OnLeave
        console.log(retValue.field<Il2Cpp.Object>("response").value.field("text").value.toString());
        console.log(retValue.field<Il2Cpp.Object>("response").value.field("data").value.toString());
        return retValue;
    }

    // Torappu.Network.WebHttpResult SendPost(System.String url, System.String param, System.String contentType); // 0x023680fa
    sendPost.implementation = function(url: Il2Cpp.String, param: Il2Cpp.String, contentType: Il2Cpp.String) : Il2Cpp.Object {
        // OnEnter
        console.log("sendPost");
        console.log(url.toString());
        console.log(param.toString());
        console.log(contentType.toString());
        const retValue = this.method<Il2Cpp.Object>("SendPost").invoke(url, param, contentType);
        // OnLeave
        console.log(retValue.field<Il2Cpp.Object>("response").value.field("text").value.toString());
        console.log(retValue.field<Il2Cpp.Object>("response").value.field("data").value.toString());
        return retValue as Il2Cpp.Object;
    }

    
    // Torappu.Network.WebHttpResult SendPost(System.String url, System.String param, System.String contentType, System.Collections.Generic.Dictionary<System.String,System.String> header); // 0x0236830a
    sendPost2.implementation = function(url: Il2Cpp.String, param: Il2Cpp.String, contentType: Il2Cpp.String, header: Il2Cpp.Object) : Il2Cpp.Object {
        // OnEnter
        console.log("sendPost2");
        console.log(url.toString());
        console.log(param.toString());
        console.log(contentType.toString());
        console.log(header.toString());
        const retValue = this.method<Il2Cpp.Object>("SendPost").overload("System.String", "System.String", "System.String", "System.Collections.Generic.Dictionary<System.String,System.String>").invoke(url, param, contentType, header);
        // OnLeave
        console.log(retValue.field<Il2Cpp.Object>("response").value.field("text").value.toString());
        console.log(retValue.field<Il2Cpp.Object>("response").value.field("data").value.toString());
        return retValue as Il2Cpp.Object;
    }
});
```

And here is the output in cmd. As I said, it just stopped after I invoke the sendGet function:
```
Spawning com.xxx.xxx...
Chrome Inspector server listening on port 9229

Spawned com.xxx.xxx. Resuming main thread!
[Android Emulator 5554::com.xxx.xxx ]-> sendGet
"https://xxx.xxx.com/xxx?sign=xxx"
null
```

**Comment by commonuserlol** [comment]
when you know return value is `System.String` (or field), you should replace call `toString()` with casting to `Il2Cpp.String` and getting `content` property. Like `console.log((clazz.method("method").invoke() as Il2Cpp.String).content)`

**Reply by lkwargs** [reply]
Thank you so much. But the app is stucked in the second sendGet function, which returns a object(Torappu.Network.WebHttpResult). How can I return that object properly?

**Reply by commonuserlol** [reply]
what do you mean "return object properly"?

**Reply by lkwargs** [reply]
I mean if this code can return a C-sharp object successfully?
```
// Torappu.Network.WebHttpResult SendGet(System.String url, System.String param); // 0x02367f2a
    sendGet.implementation = function(url: Il2Cpp.String, param: Il2Cpp.String) : Il2Cpp.Object {
        // OnEnter
        const retValue = this.method<Il2Cpp.Object>("SendGet").invoke(url, param);
        // OnLeave
        return retValue;
    }
```
I don't think the app stopped beacause of toString().

**Reply by commonuserlol** [reply]
yeah it can. if your hook don't got any error

**Comment by vfsfitvnm** [comment]
hmm I don't see anything wrong at first sight.
Perhaps you could get more insights doing:
```ts
Il2Cpp.perform(() => {
    Il2Cpp.installExceptionListener("all");
   
    ...
});
```

**Comment by lkwargs** [comment]
Thank you all guys! I solved my problem. The reason is calling `toString()` or other methods on some fields of a class that are null. It won't raise any exception, and will cause the app stopped. I think this discussion can be closed.

**Reply by vfsfitvnm** [reply]
Glad you figured out, however, is there something `frida-il2cpp-bridge` might be lacking of? E.g. something that would have prevented headache for us all, lol

**Reply by commonuserlol** [reply]
i think we miss coroutines. Ik that i should get `System.IEnumerator` from coroutine method and then invoke `MoveNext`. Tbh idk how it works so i ignore return value from `MoveNext` :/

**Comment by lkwargs** [comment]
> Glad you figured out, however, is there something `frida-il2cpp-bridge` might be lacking of? E.g. something that would have prevented headache for us all, lol

I think `frida-il2cpp-bridge` is an simple but powerful tool. However, I think it still lacks a startup document like [this](https://github.com/oleavr/frida-agent-example.git). In my learning experience, most of my time was spended on learning how to install, compile and run it. The `tsconfig.json` in wiki is out dated and causes lots of compile errors. I solved it after I find something like `"files": [ "index.ts" ]` is needed.

It's true that these questions are not related to this project, and these questions may not even be questions for those who learned js/ts before. But as I said, it's a very simple and easy tool. I have no experience on js/ts, but I can still learn using this tool to hook an app in a few hours. So I think all the people who have program skills can master this tool quickly if you give them a startup tutorials.

**Reply by vfsfitvnm** [reply]
Thanks for the precious feedback.
I'll try my best in the next months to overcome the [curse of knowledge](https://en.wikipedia.org/wiki/Curse_of_knowledge)!
